### ISSUE TITLE
Memory usage with parallel evaluation

### ISSUE THREAD

**Issue by mdewing** [stevengj/cubature#9]
For large integrals and using parallel evaluation, the memory usage for the array of regions to evaluate becomes a limiting factor, before the size of the heap becomes an issue.
Would imposing a maximum batch size maintain correctness of the parallel algorithm?
That is, break out of the inner 'do' loop in the parallel branch of rulecubature if nR exceeds a fixed size.

It seems to work empirically in a few cases, but I'm unsure of the theory.

**Comment by stevengj** [comment]
Yes, it will still be correct with a maximum batch size.  (The serial case is basically just a maximum batch size of 1.)

**Comment by loliverhennigh** [comment]
Is this mini batching implemented? This would be extremely helpful for me. Great library by the way!

**Comment by stevengj** [comment]
No, it's not currently implemented, but it would be trivial to do.   Just set a maximum number of iterations for this loop: 

https://github.com/stevengj/cubature/blob/6bda3b267b5f5d644baac09565af661487b9a0da/hcubature.c#L983-L996

For example, change the final line to:
```c
} while (regions.n > 0 && (numEval < maxEval || !maxEval) && nR < 100)
```
to set an upper bound of 100 sub-regions per iteration.

**Comment by loliverhennigh** [comment]
Neat, I will look at this. Right now my solution is to break up the batch in my given integration function. I am trying to integrate the output of a neural network and the evaluations are costly memory wise.

**Comment by mdewing** [comment]
I've implemented the maximum batch size in a branch here

https://github.com/markdewing/cubature-1/tree/max_batch_size
